sycl/include/CL/sycl/atomic.hpp

```diff
@@ -194,10 +194,32 @@ class atomic {
         Ptr, SpirvScope, detail::getSPIRVMemorySemanticsMask(Order), Operand);
   }
 
+#ifdef __SYCL_DEVICE_ONLY__
+  template <typename T2 = T>
+  detail::enable_if_t<!std::is_same<float, T2>::value, T>
+  load(memory_order Order = memory_order::relaxed) const {
+    return __spirv_AtomicLoad(Ptr, SpirvScope,
+                              detail::getSPIRVMemorySemanticsMask(Order));
+  }
+  template <typename T2 = T>
+  detail::enable_if_t<std::is_same<float, T2>::value, T>
+  load(memory_order Order = memory_order::relaxed) const {
+    auto *TmpPtr =
+        reinterpret_cast<typename multi_ptr<int, addressSpace>::pointer_t>(Ptr);
+    union {
+      int Int;
+      float Float;
+    } TmpVal;
+    TmpVal.Int = __spirv_AtomicLoad(TmpPtr, SpirvScope,
+                                    detail::getSPIRVMemorySemanticsMask(Order));
+    return TmpVal.Float;
```

Comment:
Doing a bitcast by reading from an non-active member of an union is UB in C++. Might be better to use __builtin_bitcast.

Reply:
> Might be better to use __builtin_bitcast.

Does it supported by MSVC? or by GCC? Probably it is better to go with `memcpy` if we want to avoid UB here (which is, however, works fine on most compilers)

Reply:
No this is clang only. But the code is inside `#ifdef__SYCL_DEVICE_ONLY__` . And my understanding is that device only code only needs to support clang. I agree that using `memcpy` is the correct solution if it needs to be portable.

Reply:
Is it better to use `reinterpret_cast` instead of `memcpy`?

Reply:
`reinterpret_cast`  is UB too. If the code is supposed to not be UB and be portable `memcpy` is the only option.  If we consider UB to be fine as long as the compilers we care about treat it as implementation defined, either union or `reinterepret_cast` could be fine. But I don't know whether MSVC makes any guarantees for either union or `reinterepret_cast`.